File: JavaScript/094_inorderTraversal.js

// Definition for a binary tree node.
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

let node1 = new TreeNode(1);
let node2 = new TreeNode(2);
let node3 = new TreeNode(3);
node1.right = node2;
node2.left = node3;

// /**
//  * @param {TreeNode} root
//  * @return {number[]}
//  */
// var inorderTraversal = function(root) {
//   if (!root) return [];
//   const result = [];
//   if (root.left) {
//     result.push(...inorderTraversal(root.left))
//   }
//   result.push(root.val)
//   if (root.right) {
//     result.push(...inorderTraversal(root.right))
//   }
//   return result;
// };

// /**
//  * @param {TreeNode} root
//  * @return {number[]}
//  */
// var inorderTraversal = function(root) {
//   let result = [];
//   dfs(root);

//   function dfs(root) {
//     if (root) {
//       dfs(root.left)
//       result.push(root.val)
//       dfs(root.right)
//     }
//   }

//   return result;
// };

/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
  let stack = [];
  let result = [];
  let current = root;
  while (stack.length > 0 || current) {
    while (current) {
      stack.push(current);
      current = current.left;
    }
    current = stack.pop();
    console.log(current)
    result.push(current.val);
    current = current.right;
  }
  return result;
};

console.log(inorderTraversal(node1));

File: Python/094_inorderTraversal.py
from typing import List, Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution1:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        if root:
            res = self.inorderTraversal(root.left)
            res.append(root.val)
            res = res + self.inorderTraversal(root.right)
        return res

class Solution2:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        stack = []
        result = []
        curr = root
        while stack or curr:
            # Traverse in inorder way(left node first)
            while curr:
                stack.append(curr)
                curr = curr.left
            # Back to the parent node and traverse to right node
            curr = stack.pop()
            result.append(curr.val)
            curr = curr.right

        return result


if __name__ == '__main__':
    root = TreeNode(1)
    node2 = TreeNode(2)
    node3 = TreeNode(3)
    root.right = node2
    node2.left = node3
    sol = Solution2()
    print(sol.inorderTraversal(root))